File: tests/ts/smoke.test.ts
import { readFileSync } from "node:fs";
import { join, dirname } from "node:path";
import { fileURLToPath } from "node:url";
import { Convert } from "../../generated/ts/src/types.js";

const __dirname = dirname(fileURLToPath(import.meta.url));
const examplesDir = join(__dirname, "..", "..", "examples");

function readExample(path: string): string {
  return readFileSync(join(examplesDir, path), "utf-8");
}

let failures = 0;

function assert(condition: boolean, msg: string): void {
  if (!condition) {
    console.error(`FAIL: ${msg}`);
    failures++;
  }
}

// Settings round-trip
const settingsJson = readExample("settings/valid/minimal.json");
const settings = Convert.toClaudeCodeSettings(settingsJson);
const settingsOut = Convert.claudeCodeSettingsToJson(settings);
const settingsParsed = JSON.parse(settingsOut);
assert(
  settingsParsed.permissions?.defaultMode === "default",
  "settings round-trip: defaultMode should be 'default'"
);

// Plugin manifest round-trip
const pluginJson = readExample("plugins/valid/full-metadata.json");
const plugin = Convert.toClaudeCodePluginManifest(pluginJson);
const pluginOut = Convert.claudeCodePluginManifestToJson(plugin);
const pluginParsed = JSON.parse(pluginOut);
assert(
  pluginParsed.name === "code-review-helper",
  "plugin round-trip: name should be 'code-review-helper'"
);
assert(
  pluginParsed.version === "1.2.0",
  "plugin round-trip: version should be '1.2.0'"
);
assert(
  pluginParsed.license === "MIT",
  "plugin round-trip: license should be 'MIT'"
);

// Permissions example
const permJson = readExample("settings/valid/permissions-full.json");
const perm = Convert.toClaudeCodeSettings(permJson);
const permOut = Convert.claudeCodeSettingsToJson(perm);
const permParsed = JSON.parse(permOut);
assert(
  Array.isArray(permParsed.permissions?.allow),
  "permissions: allow should be an array"
);
assert(
  Array.isArray(permParsed.permissions?.deny),
  "permissions: deny should be an array"
);

if (failures > 0) {
  console.error(`\n${failures} assertion(s) failed`);
  process.exit(1);
} else {
  console.log("All smoke tests passed");
}

File: tests/go/smoke_test.go
package claudeconfig_test

import (
	"os"
	"testing"

	claudeconfig "github.com/phatblat/claude-config-schema/generated/go"
)

func TestSettingsUnmarshalMarshal(t *testing.T) {
	data, err := os.ReadFile("../../examples/settings/valid/minimal.json")
	if err != nil {
		t.Fatalf("failed to read example: %v", err)
	}

	settings, err := claudeconfig.UnmarshalClaudeCodeSettings(data)
	if err != nil {
		t.Fatalf("failed to unmarshal: %v", err)
	}

	if settings.Permissions == nil {
		t.Fatal("expected permissions to be non-nil")
	}

	if settings.Permissions.DefaultMode == nil || string(*settings.Permissions.DefaultMode) != "default" {
		t.Errorf("expected defaultMode to be 'default'")
	}

	_, err = settings.Marshal()
	if err != nil {
		t.Fatalf("failed to marshal: %v", err)
	}
}

func TestPluginManifestUnmarshalMarshal(t *testing.T) {
	data, err := os.ReadFile("../../examples/plugins/valid/full-metadata.json")
	if err != nil {
		t.Fatalf("failed to read example: %v", err)
	}

	manifest, err := claudeconfig.UnmarshalClaudeCodePluginManifest(data)
	if err != nil {
		t.Fatalf("failed to unmarshal: %v", err)
	}

	if manifest.Name != "code-review-helper" {
		t.Errorf("expected name 'code-review-helper', got %v", manifest.Name)
	}

	if manifest.Version == nil || *manifest.Version != "1.2.0" {
		t.Errorf("expected version '1.2.0', got %v", manifest.Version)
	}

	if manifest.License == nil || *manifest.License != "MIT" {
		t.Errorf("expected license 'MIT', got %v", manifest.License)
	}

	_, err = manifest.Marshal()
	if err != nil {
		t.Fatalf("failed to marshal: %v", err)
	}
}

func TestSettingsPermissions(t *testing.T) {
	data, err := os.ReadFile("../../examples/settings/valid/permissions-full.json")
	if err != nil {
		t.Fatalf("failed to read example: %v", err)
	}

	settings, err := claudeconfig.UnmarshalClaudeCodeSettings(data)
	if err != nil {
		t.Fatalf("failed to unmarshal permissions example: %v", err)
	}

	if settings.Permissions == nil {
		t.Fatal("expected permissions to be non-nil")
	}

	if len(settings.Permissions.Allow) == 0 {
		t.Error("expected allow rules to be non-empty")
	}

	if len(settings.Permissions.Deny) == 0 {
		t.Error("expected deny rules to be non-empty")
	}
}
